server/src/main/java/de/uftos/repositories/solver/timefold/constraints/constraintinstances/ConstraintInstanceHardPenalize.java

```diff
@@ -0,0 +1,38 @@
+package de.uftos.repositories.solver.timefold.constraints.constraintinstances;
+
+import de.uftos.dto.solver.RewardPenalize;
+import de.uftos.repositories.solver.timefold.domain.ResourceTimefoldInstance;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * This class models a constraint instance, which can be used by the Timefold solver.
+ * This instance is hard penalized when the evaluation function returns true.
+ *
+ * @param parameters         the resources used in this constraint instance.
+ * @param evaluationFunction the function used to evaluate this constraint instance.
+ */
+public record ConstraintInstanceHardPenalize(
+    List<ResourceTimefoldInstance> parameters,
+    Function<List<ResourceTimefoldInstance>, Boolean> evaluationFunction
+) implements ConstraintInstanceTimefoldInstance {
+  @Override
+  public boolean evaluate(HashMap<String, ResourceTimefoldInstance> resources) {
+    List<ResourceTimefoldInstance> params = new ArrayList<>();
+    for (ResourceTimefoldInstance resource : parameters) {
+      params.add(resources.get(resource.getId()));
+    }
+    try {
+      return evaluationFunction.apply(params);
+    } catch (NullPointerException e) {
+      return true;
+    }
```

Comment:
Catching a null pointer exception here look sketchy. Why do you need that? You should much rather just do `if (something == null) { return true; }`

Reply:
It looks sketchy, because it was...
Initially the Constraint-Provider (or EasyScoreCalculator) would fill the null-pointers of the lessons in, however they would check the constraints after each null-pointer got set and therefore those constraints would all throw null-pointer exceptions.
And it wasn't possible to check the relevant resources beforehand, as those relevant resources were unknown.
However by now the null values get pre-instantiated by the Repository instead of the Solver, so i should be able to simply remove these.

Reply:
Nice!